File: ts/project/.sdk/tm/rust/utility/done.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::error::ProjectNameError;
use crate::core::helpers::{getp, to_map};
use crate::utility::voxgigstruct as vs;
use crate::utility::voxgigstruct::Value;

pub fn done_util(ctx: &Rc<Context>) -> Result<Value, ProjectNameError> {
    {
        let ctrl = ctx.ctrl.borrow().clone();
        let mut c = ctrl.borrow_mut();
        if c.has_explain() {
            let explain = crate::utility::clean::clean_util(ctx, &c.explain);
            if let Value::Map(_) = getp(&explain, "result") {
                let rm = to_map(&getp(&explain, "result"));
                vs::del_prop(rm, &Value::str("err"));
            }
            c.explain = explain;
        }
    }

    let result = ctx.result.borrow().clone();
    if let Some(res) = result {
        if res.borrow().ok {
            return Ok(res.borrow().resdata.clone());
        }
    }

    crate::utility::make_error::make_error_util(ctx, None)
}

File: ts/project/.sdk/tm/java/utility/Done.java
package JAVAPACKAGE.utility;

import java.util.Map;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.Helpers;

@SuppressWarnings({"unchecked"})
final class Done {

  private Done() {}

  static Object done(Context ctx) {
    if (ctx.ctrl.explain != null) {
      ctx.ctrl.explain = (Map<String, Object>) Clean.clean(ctx, ctx.ctrl.explain);
      Object explainResult = ctx.ctrl.explain.get("result");
      Map<String, Object> rm = Helpers.toMapAny(explainResult);
      if (rm != null) {
        rm.remove("err");
      }
    }

    if (ctx.result != null && ctx.result.ok) {
      return ctx.result.resdata;
    }

    return MakeError.makeError(ctx, null);
  }
}
